.circleci/config.yml

```diff
@@ -26,107 +34,237 @@ executors:
+  ######################
+  # common setup steps #
+  ######################
+
+  install-deps-yum:
+    parameters:
+      bazel-arch:
+        type: string
+    steps:
+      - run: |
+          amazon-linux-extras install python3.8 java-openjdk11 -y
+          yum install -y git tar gcc gcc-c++ file lsof which procps llvm-devel clang-devel
+          curl -OL "https://github.com/bazelbuild/bazelisk/releases/download/v1.17.0/bazelisk-linux-<<parameters.bazel-arch>>"
+          mv "bazelisk-linux-<<parameters.bazel-arch>>" /usr/local/bin/bazel
+          chmod a+x /usr/local/bin/bazel
+
+  install-deps-apt:
+    parameters:
+      bazel-arch:
+        type: string
     steps:
       - run: |
-          curl -OL "https://github.com/bazelbuild/bazelisk/releases/download/v1.17.0/bazelisk-linux-amd64"
-          sudo mv "bazelisk-linux-amd64" /usr/local/bin/bazel
+          apt update -y
+          DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends tzdata
+          apt install -y curl build-essential git python3 python3-pip default-jre lsof cmake file wget libclang-dev
+          curl -OL "https://github.com/bazelbuild/bazelisk/releases/download/v1.17.0/bazelisk-linux-<<parameters.bazel-arch>>"
+          mv "bazelisk-linux-<<parameters.bazel-arch>>" /usr/local/bin/bazel
           chmod a+x /usr/local/bin/bazel
 
-  install-bazel-linux-arm64:
+  install-deps-brew:
+    parameters:
+      bazel-arch:
+        type: string
     steps:
       - run: |
-          curl -OL "https://github.com/bazelbuild/bazelisk/releases/download/v1.17.0/bazelisk-linux-arm64"
-          sudo mv "bazelisk-linux-arm64" /usr/local/bin/bazel
+          brew install python@3.8
+          curl -OL "https://github.com/bazelbuild/bazelisk/releases/download/v1.17.0/bazelisk-darwin-<<parameters.bazel-arch>>"
+          sudo mv "bazelisk-darwin-<<parameters.bazel-arch>>" /usr/local/bin/bazel
           chmod a+x /usr/local/bin/bazel
 
-  install-bazel-mac:
+  install-brew-rosetta:
     steps:
-      - run: brew install bazelisk
+      - run: arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+
+
+  # deploy & test steps
+  test-assembly-unix:
+    steps:
+      - run: |
+          bazel build //:assemble-typedb-server
+          bazel test //tests/assembly:test_assembly --test_output=errors
+
+  deploy-snapshot-unix:
+    steps:
+      - run: |
+          export DEPLOY_ARTIFACT_USERNAME=$REPO_TYPEDB_USERNAME
+          export DEPLOY_ARTIFACT_PASSWORD=$REPO_TYPEDB_PASSWORD
+          bazel run --jobs=8 --define version=$(git rev-parse HEAD) //:deploy-typedb-server --compilation_mode=opt -- snapshot
+
+  deploy-release-unix:
+    steps:
+      - run: |
+          export DEPLOY_ARTIFACT_USERNAME=$REPO_TYPEDB_USERNAME
+          export DEPLOY_ARTIFACT_PASSWORD=$REPO_TYPEDB_PASSWORD
+          bazel run --jobs=8 --define version=$(cat VERSION) //:deploy-typedb-server --compilation_mode=opt -- release
```

Comment:
Can you try figuring out how to use `lto` optimisations (link time optimisations) in rust too? Hopefully we can pass it through bazel somehow :O maybe test it locally to see what it does to compile time first...
If possible we should add it to the release jobs. If it's really not different we can add it to snapshot deployment jobs too!